nameko_stripe/handlers/webhook.py

```diff
@@ -0,0 +1,73 @@
+import logging
+
+import stripe
+from nameko.exceptions import ConfigurationError
+from nameko.web.handlers import HttpRequestHandler
+from werkzeug.exceptions import BadRequest, HTTPException, Unauthorized
+
+
+from nameko_stripe import constants
+
+
+logger = logging.getLogger(__name__)
+
+
+class WebhookHandler(HttpRequestHandler):
+    def __init__(self, path=None, api_key=None, endpoint_secret=None, **kwargs):
+        self.api_key = api_key
+        self.endpoint_secret = endpoint_secret
+        super(WebhookHandler, self).__init__("POST", path, **kwargs)
+
+    def setup(self):
+
+        config = self.container.config.get(constants.CONFIG_KEY, {})
+
+        try:
+            self.api_key = self.api_key or config["SECRET_KEY"]
+            self.endpoint_secret = self.endpoint_secret or config["ENDPOINT_SECRET"]
+        except KeyError as exc:
+            raise ConfigurationError(
+                "Please provide {} for stripe API communication".format(exc.args[0])
+            ) from exc
+
+        if not self.url:
+            self.url = config.get("DEFAULT_PATH", constants.DEFAULT_EVENT_PATH)
+
+        super(WebhookHandler, self).setup()
+
+    def parse_event(self, request):
+        try:
+            return stripe.Webhook.construct_event(
```

Comment:
There is no client used here, no setup neither app info needed. Here is just the incoming event parsed and sender authenticated using the stripe lib.

Reply:
The secrets here are used to check the payload signature provided in the auth header.

Reply:
I see your point.

My suggestion was to have a `self.stripe = stripe` or `self.client = stripe` and do `self.client.set_app_info(**constants.APP_INFO)` as stated [here](https://github.com/stripe/stripe-python#writing-a-plugin) the same way that we do in the dependency provider. So, in case this module is going to perform any calls to the Stripe API, this info is also passed. https://github.com/stripe/stripe-python#writing-a-plugin

I'll leave this up to you.

Reply:
Thanks for the suggestion @juliotrigo! As I wrote earlier there is no use of the client and no app info is sent, in fact there is nothing sent over the wire. What you are suggesting is to take a whole package and place it on an instance attribute which I think is wrong. Doing `self.stripe = stripe` is like doing  `self.nameko = nameko` or `self.django = django`. I don't think it is wise doing, instead you just import the module where you need it and use it straight away.

I see your point of setting things up once and on one place. If you have a look of what the function does, it set a _global_ app info which is then used by any call. It's sort of fine as there would be most probably only one app used by the plugin.

With `api_key` it's different story though as you can have multiple platform stripe accounts used in one service so secrets should not be set globally I think. One of common cases is using different account for different currencies to avoid expensive currency conversions. We do it in our payment service for example (we had recommended this approach by Stripe devs and Stripe product guys). So instead of setting API key globally for all clients we pass secrets with each call to the API (done by provider, not inside entrypoints). The Stripe python package does not provide an API based on client instances but it allows for passing secrets extra for each call and the global setup is used only as convenient default for simple one-account cases (the [link you are referring to](https://github.com/stripe/stripe-python#writing-a-plugin) is a recommendation for such simple case)

The current implementation of the dependency provider sets things globally and does not allow for multi-account implementations. _Also it does exactly the thing with placing the whole module on the provider instance which is then injected to be used byt the entrypoint_.

I think that better would be to return a wrapper which wraps the underlying API and passes secrets loaded in init and setup to each underlying stripe resource method.

Following is an example of a manually wrapped method (can be done automatically):

```
    @idempotent_retry(
        for_exceptions=stripe.error.APIConnectionError,
        delay=1,
        max_attempts=5)
    def create_payout(self, **kwargs):
        """ Create a payout object
        :returns:
            A dictionary representing the created Payout object.
        """
        return stripe.Payout.create(api_key=self.api_key, **kwargs)
```

This example is taken from our client wrapper which is returned by our stripe dependency provider. It normally uses `stripe` module (there is no `self.stripe`) and it also demonstrate additional benefits of wrapping stripe package API instead of returning the package - idempotent retry implemented on stripe recommendation you can find int [Stripe docs](https://stripe.com/docs/api/idempotent_requests) (one of the tooling I would also like to add to this lib).

This is a bit bigger topic then, outside of this PR ;) I'm placing this here as my input and suggestion for the future development of this nameko extension. As a conclusion I think that a) we should not place whole package/module on extension instance attr and b) we should not configure stripe lib globally.

Instead we should thinking of rewriting the dependency provider to return a wrapper suggested above which a) leaves normal lib package usage on service developers (import and use).
b) supports the use of multiple stripe accounts in one service c) encourages for keeping underlying lib implementation details inside providers

Reply:
Thanks for the detailed explanation @iky ! Much appreciated.

Also, really interesting to read how you use a different account for different currencies to avoid expensive currency conversions. We will take this into consideration.

This is a very interesting discussion, thangs for bring it up. Even if it does not belong to this PR, it may help us think about how we want to move forward.

Regarding the way `self.stripe = stripe` is currently done, what you're saying makes completely sense. We have to consider that a) the inital intention was to have a very minimalistic dependency provider that could be extended by anyone to add more specific logic b) we did not have the use case of multiple accounts per service/dependency. The current version sets the secret (just once) and returns an object that can easily be used by the service (or another dependency that extends this one).

That being set, by all means, we should improve it and make it more extensible. I have a few comments/question regarding what you said:

* In our codebase, we use a dependency provider that extends this one and returns a wrapper, as you say (we do not deal with Stripe logic outside a dependency provider). We do not need to set the `api_key` though, as it is already set by this dependency, but we have specific methods for the different pieces of functionality that we need, like the one that you posted above (without the decorator though). By adding all these specific methods, we thought that we could be adding too much logic into this library and there's sometimes a fine line between something generic vs. something slightly coupled to the specific domain of your application. Maybe other users want to code the individual stripe calls independently.

* We can always have the dependency return a wrapper, that can be easily extensible or replaceable.

```python
class StripeWrapper():

    def __init__(self, api_key=None):
        self.api_key = api_key

    # The wrapper that you mention


class Stripe(DependencyProvider):

    stripe_wrapper_cls = StripeWrapper
    
    # ...

    def get_dependency(self, worker_ctx):
        return stripe_wrapper_cls(self.api_key)
```
However, this raises the question of how much logic we put into the wrapper. You showed the `stripe.Payout.create` case and we have a dozen of other similar methods in our codebase (`client.Customer.create`, `SKU.retrieve`, `Order.retrieve`), but should we add all those methods here or do they belong to the specific codebases that use this dependency? How far should we go and how much logic should we put into this dependency/wrapper? Since we are not going to add methods for all the logic that Stripe supports, we'll end up with a (big) wrapper that just supports a few of them. For the rest, you'd have to extend the wrapper or use your own wrapper. Maybe this is something that could be left to the users to implement themselves according to their needs?

I guess one of the main points is how much logic we add to this library and how we do it so that we keep it simple and extensible by other users.

I would like to know what your view is on these points.

Reply:
It is a good point @juliotrigo! how far we should go with the wrapper?

What we can do is to proxy attributes and methods automatically similarly as it is done in [nameko-salesforce](https://github.com/nameko/nameko-salesforce/tree/master/nameko_salesforce/api). The dependency provider can return a similar wrapper client which automatically proxies [attributes](https://github.com/nameko/nameko-salesforce/blob/master/nameko_salesforce/api/client.py#L66) (resources such as `Source` or `Charge` in our stripe case) and their [methods](https://github.com/nameko/nameko-salesforce/blob/master/nameko_salesforce/api/client.py#L27) (such as `get` or `create`) to the underlying stripe package.

This way we do not have to prepare any of the methods manually neither we have to maintain them = lightweight codebase always supporting the whole and latest stripe lib API. This might be quite elegant way of dealing with the issue, what do you think?